vertx-web-common/src/main/java/io/vertx/ext/web/codec/sse/SseBodyCodec.java

```diff
@@ -0,0 +1,248 @@
+      // Field names must be compared literally, with no case folding performed.
+      switch (field) {
+        case "event":
+          builder.event(value);
+          break;
+        case "data":
+          builder.data(value);
+          break;
+        case "id":
+          builder.id(value);
+          break;
+        case "retry":
+          // If the field value consists of only ASCII digits, then interpret the field value as an
+          // integer in base ten, and set the event stream's reconnection time to that integer.
+          // Otherwise, ignore the field.
+          try {
+            builder.retry(Integer.parseInt(value));
+          } catch (NumberFormatException ex) {
+            throw new RuntimeException("Invalid \"retry\" value:" + value, ex);
+          }
+          break;
+        default:
+          // Ignore unknown fields as per SSE spec
+          break;
+      }
+    }
+
+    synchronized SseEvent nextSseEvent() {
+      SseEvent.Builder eventBuilder = SseEvent.builder();
+      int lineStart = 0;
+      byte[] bytes = content.getBytes();
+
+      for (int i = 0; i < bytes.length; i++) {
+        byte b = bytes[i];
+        if (b == '\n' || b == '\r') {
+          // Extract the line without the newline character
+          String line = content.getString(lineStart, i, "UTF-8");
+
+          if (line.isEmpty()) {
+            // Empty line dispatches the event
+            content = content.getBuffer(i + 1, content.length());
+            return eventBuilder.build();
+          } else {
+            parseLine(line, eventBuilder);
+          }
+
+          lineStart = i + 1;
+        }
+      }
+      return null;
+    }
+
+    void check() {
+      if (failed) {
+        return;
+      }
+      while (true) {
+        long d = demand.get();
+        if (d == 0L) {
+          break;
+        }
+        SseEvent event;
+        try {
+          event = nextSseEvent();
+        } catch (Exception e) {
+          failed = true;
+          handle(e);
+          handleEnd();
+          return;
+        }
+        if (event == null) {
+          if (ended) {
+            handleEnd();
+          }
+          break;
+        }
+        // Race possible
+        if (d != Long.MAX_VALUE) {
+          demand.decrementAndGet();
+        }
+        Handler<SseEvent> h = handler;
+        if (h != null) {
+          h.handle(event);
+        }
+      }
+      if (content.length() < LOW_WATERMARK && writeQueueFull) {
+        writeQueueFull = false;
+        Handler<Void> h = drainHandler;
+        if (h != null) {
+          h.handle(null);
+        }
+      }
+    }
+
+    private void handleEnd() {
+      Handler<Void> h = endHandler;
+      if (h != null) {
+        h.handle(null);
+      }
+    }
+    @Override
+    public Future<Void> write(Buffer buffer) {
+      content.appendBuffer(buffer);
+      check();
+      writeQueueFull |= writeQueueFull();
```

Comment:
I think check should return whether the queue is full or not and use the result from check(), this avoids a possible race to do the update in check and the read the value, while check can compute it correctly